lib/iterator.js

```diff
@@ -0,0 +1,312 @@
+        return value;
+      }
+    }
+  }
+
+  includes(element) {
+    for (const value of this) {
+      if (value === element) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  reduce(reducer, initialValue) {
+    let result = initialValue;
+
+    if (result === undefined) {
+      const next = this.next();
+      if (next.done) {
+        throw new TypeError(
+          'reduce of consumed iterator with no initial value'
+        );
+      }
+      result = next.value;
+    }
+
+    for (const value of this) {
+      result = reducer(result, value);
+    }
+    return result;
+  }
+
+  some(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  collectTo(CollectionClass) {
+    return new CollectionClass(this);
+  }
+
+  collectWith(obj, collector) {
+    this.forEach(element => collector(element, obj));
+  }
+
+  toArray() {
+    return Array.from(this);
+  }
+
+  map(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new MapIterator(this, mapper, thisArg);
+  }
+
+  filter(predicate, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FilterIterator(this, predicate, thisArg);
+  }
+
+  flat(depth = 1) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatIterator(this, depth);
+  }
+
+  flatMap(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatMapIterator(this, mapper, thisArg);
+  }
+
+  zip(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new ZipIterator(this, iterators);
+  }
+
+  join(...iterators) {
+    // eslint-disable-next-line no-use-before-define
+    return new JoinIterator(this, new Iterator(iterators));
+  }
+
+  take(amount) {
+    // eslint-disable-next-line no-use-before-define
+    return new TakeIterator(this, amount);
+  }
+
+  skip(amount) {
+    // eslint-disable-next-line no-use-before-define
+    for (let i = 0; i < amount; i++) {
+      this.next();
+    }
+    return this;
+  }
+}
+
+class MapIterator extends Iterator {
+  constructor(base, mapper, thisArg) {
+    super(base);
+    this.mapper = mapper;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    const { done, value } = this.base.next();
+    return {
+      done,
+      value: done ? undefined : this.mapper.call(this.thisArg, value),
+    };
+  }
+}
+
+class FilterIterator extends Iterator {
+  constructor(base, predicate, thisArg) {
+    super(base);
+    this.predicate = predicate;
+    this.thisArg = thisArg;
+  }
+
+  next() {
+    for (const value of this.base) {
+      if (this.predicate.call(this.thisArg, value)) {
+        return { done: false, value };
+      }
+    }
+    return { done: true, value: undefined };
+  }
+}
+
+class FlatIterator extends Iterator {
+  constructor(base, depth) {
+    super(base);
+    this.depth = depth;
+    this.stack = [base];
```

Comment:
You know the maximum stack size at the time of creation, so you may as well just create a fixed size array here, to avoid additional memory allocations.

Reply:
I wouldn`t like to do so, because then I'll need to store the current index. Also, this looks like premature optimisations. You can change that in a follow-up PR.

Reply:
You don't need to store current index, you can use `depth` field for that.

Reply:
I need to know both desired depth and current depth, that's why I'll need an additional index.

Reply:
Yea, I see now, I will not need a `depth` variable as desired depth can be inferred from `this.stack.length`.